test/distributed/org/apache/cassandra/distributed/shared/WithProperties.java

```diff
@@ -0,0 +1,114 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.cassandra.distributed.shared;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+
+import com.google.common.base.Joiner;
+
+import org.apache.cassandra.config.CassandraRelevantProperties;
+
+public final class WithProperties implements AutoCloseable
+{
+    private final List<State> properties = new ArrayList<>();
+
+    public WithProperties()
+    {
+    }
+
+    public WithProperties(String... kvs)
+    {
+        with(kvs);
+    }
+
+    public void with(String... kvs)
+    {
+        assert kvs.length % 2 == 0 : "Input must have an even amount of inputs but given " + kvs.length;
+        for (int i = 0; i <= kvs.length - 2; i = i + 2)
+        {
+            with(kvs[i], kvs[i + 1]);
+        }
+    }
+
+    public void setProperty(String key, String value)
+    {
+        with(key, value);
+    }
+
+    public void set(CassandraRelevantProperties prop, String value)
+    {
+        with(prop.getKey(), value);
+    }
+
+    public void set(CassandraRelevantProperties prop, String... values)
+    {
+        set(prop, Arrays.asList(values));
+    }
+
+    public void set(CassandraRelevantProperties prop, Collection<String> values)
+    {
+        set(prop, Joiner.on(",").join(values));
+    }
+
+    public void set(CassandraRelevantProperties prop, boolean value)
+    {
+        set(prop, Boolean.toString(value));
+    }
+
+    public void set(CassandraRelevantProperties prop, long value)
+    {
+        set(prop, Long.toString(value));
+    }
+
+    public void with(String key, String value)
+    {
+        String previous = System.setProperty(key, value);
+        properties.add(new State(key, previous));
+    }
+
+
+    @Override
+    public void close()
+    {
+        Collections.reverse(properties);
+        properties.forEach(s -> {
+            if (s.value == null)
+                System.getProperties().remove(s.key);
+            else
+                System.setProperty(s.key, s.value);
+        });
+        properties.clear();
+    }
+
+    private static final class State
```

Comment:
I suck at naming and named it what it is, its the state of this data structure =D

though, ill rename Property as it flows better in this class

> key not name?

This is because for properties they are key/value pairs, so key feels better to me than name does.

Reply:
Fine with me - I checked the `setProperty` API and that's what it calls the arguments (I thought they were property names, not keys).

Reply:
I can switch to name

Reply:
weird... IntelliJ is saying its `key` for me

```
# System
public static String setProperty(String key, String value) {
# Properties
public synchronized Object setProperty(String key, String value) {
```

https://github.com/openjdk/jdk/blob/master/src/java.base/share/classes/java/util/Properties.java#L228


I will leave it as key =D